### ISSUE TITLE
Not possible to route inbound (direction 2) on smsc_id only

### ISSUE THREAD

**Issue by dspasoski** [kneodev/ksmppd#29]
Hi,
I'm having a problem routing inbound (MO->AT) SMS using the DB method (smpp_route) using only smsc_id as a parameter and setting .* (or even null) for destination addresses. ksmppd will always match the first record having regex .* and direction 2, i.e. smsc_id is not taken into account.
Please elaborate the matching algorithm and the fields it considers.

**Comment by dspasoski** [comment]
Hi,

It looks like the patch that fixed issue #29 opened a new problem.
If I try to route based on the smpp_route table, I get the following error:

ERROR: MYSQL: Unable to prepare statement: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'AND system_id = ?' at line 1
2017-02-08 10:01:53 [16877] [10] WARNING: SMPP[testgw] could not route message to 38975400608, rejecting

If the SQL is printed, then it is the following construction:
SELECT `regex`, `cost`, `system_id`, `smsc_id`, `source_regex` FROM smpp_route WHERE direction = 1 ORDER BY priority DESC AND system_id = ?

The problem is the AND system_id afther the ORDER BY

I have the following lines in the user and route tables:
smpp_route
'18', '1', '.*', NULL, 'testgw', 'testgw', '.*', '0'

smpp_user
'testgw', '*D9FB264D42F935578D7C50EB6C04ACDAE3949774', '0.00000', NULL, '0', '0', '0', NULL, '0', '0', '0', '0', '0', '3', NULL

But generally, I can’t understand the routing logic. How do the tables interact when the smsc_id in smpp_user is <> null and when = null? Does setting smsc_id <>  null defeat the smpp_route table or they work together?

Example:

-          I want that esme to always end_up at the same smsc_id

-          I want to set some rules in the routing table that will allow only certain sources and destinations, i.e. reject everything else with some of 0x0A, 0X0B or at least 0x45
Is the above possible?

BR,

Davor Spasoski

**Comment by kneodev** [comment]
Make sure your permissions allow KSMPPD to alter tables or it will have problems. It automatically adds new columns in new versions which looks like you don't have.

**Comment by dspasoski** [comment]
Hi,

I have the required privileges and these are the columns:

Smpp_user
system_id,  password, throughput,  default_smsc,  default_cost,  enable_prepaid_billing, credit, callback_url, simulate, simulate_deliver_every, simulate_permanent_failure_every,
simulate_temporary_failure_every, simulate_mo_every, max_binds, connect_allow_ip

Smpp_route
route_id, direction, regex, cost system_id, smsc_id, source_regex, priority

It looks like the problem is with the sql in smpp_database_mysql.c . Here’s what a colleague pointed: The sql string is appended with “AND system_id” after the if. This is causing ksmppd to spit the SQL error. He did a workaround below, which solved the problem, but I wouldn’t like to apply local patches because we don’t have the full picture of how ksmppd works. If you think the patch is OK, please add it to the main release. Kindly, please drop a couple of lines of the routing decision flow.


sql = octstr_format("SELECT `regex`, `cost`, `system_id`, `smsc_id`, `source_regex` FROM %S WHERE direction = %d ORDER BY priority DESC", smpp_server->database_route_table, direction);

    if(octstr_len(service)) {
        octstr_format_append(sql, " AND system_id = ?");
        binds = gwlist_create();
        gwlist_produce(binds, service);
    }

The fix:
sql = octstr_format("SELECT `regex`, `cost`, `system_id`, `smsc_id`, `source_regex` FROM %S WHERE direction = %d", smpp_server->database_route_table, direction);

    if(octstr_len(service)) {
        octstr_format_append(sql, " AND system_id = ?");
        binds = gwlist_create();
        gwlist_produce(binds, service);
    }

    octstr_format_append(sql," ORDER BY priority DESC");

BR and nice weekend!

Davor Spasoski

**Comment by kneodev** [comment]
Apologies there was actually a bug introduced, I have patched in git

**Comment by dspasoski** [comment]
I had to modify the install script, some URLs were invalid and there was a wrong link created.

BR,

Davor Spasoski

**Comment by kneodev** [comment]
Please elaborate

**Comment by dspasoski** [comment]
1. I had to change fto to https, but this is porbably a local problem on my machine. For some reason I can’t open the ftp site. Anyway, http and https are more likely to be allowed on corporate firewalls.
2.The second change is more substantial. The script complained that it can’t cd to libevent/.libs

diff bootstrap.sh /home/vas/sources/ksmppd/bootstrap.sh
51c51
< safe_exec "curl -O -L ftp://ftp.gnu.org/gnu/shtool/shtool-2.0.8.tar.gz"
---
> safe_exec "curl https://ftp.gnu.org/gnu/shtool/shtool-2.0.8.tar.gz"
75c75
< ln -s libevent-2.0.22-stable libevent
---
> ln -s ../libevent-2.0.22-stable libevent

Davor Spasoski
From: Davor Spasoski
Sent: 14 February 2017 12:59
To: 'kneodev/ksmppd' <reply@reply.github.com>; kneodev/ksmppd <ksmppd@noreply.github.com>
Cc: Author <author@noreply.github.com>
Subject: RE: [kneodev/ksmppd] Not possible to route inbound (direction 2) on smsc_id only (#29)

I had to modify the install script, some URLs were invalid and there was a wrong link created.

BR,

Davor Spasoski
